### ISSUE TITLE
Tab completion

### ISSUE THREAD

**Issue by wernight** [lebedov/ripdb#1]
Tab completion is one of the main things I love about ipdb. Please support and document it.

**Comment by lebedov** [comment]
I would really like to support this, but it doesn't appear to be straightforward to do over a network connection (a question [I raised in the past](http://stackoverflow.com/questions/30289062/triggering-remote-programs-tab-completion-through-telnetlib-with-python-2) when I was looking into how to support tab completion still hasn't elicited a response).

**Comment by wernight** [comment]
I'm not sure if tabs are sent. Could try sending another character (anything first for testing and later may be Ctrl + N to be like VIM), and server-side to the conversion.

If the issue is in the outputs of ipdb, I'm less familiar, just would flush of course. Normally they should only be ANSI codes.

**Comment by lebedov** [comment]
Making some more inquiries to see if the Jupyter folks have any suggestions.

**Comment by wernight** [comment]
They might. Have you tried somehow faking the send of `\t` to see what happens? I'm kind of hoping you've an idea how to test that quickly.

**Comment by lebedov** [comment]
One problem is that clients like telnet do not automatically propagate control characters like tab associated with rich terminals. Based upon [this post](http://superuser.com/questions/123790/socat-and-rich-terminals-with-ctrlc-ctrlz-ctrld-propagation), I gathered that something like socat might be able to send the control chars correctly, but didn't obtain any success the last time I tried.

The Juputer folks [suggested](https://groups.google.com/forum/#!topic/jupyter/_r8bkJ_pK7k) wrapping the IPython debugger with a Jupyter kernel so that it can be embedded in a Python session and remotely accessed via the Jupyter console.

**Comment by wernight** [comment]
The Jupyter way is an idea.

What I was suggesting (outside of using socat) is to remap for example X
key to <Tab>, just for testing to see if it works. If it does, may be remap
Ctrl+N and/or Ctrl+P to Tab (those are one of the common Tab alternative
IMO), Ctrl+Space may be another good one.

If that works, just having it in the read-me should be enough.

**Comment by x746e** [comment]
I did a bit of research, and it turned out that:
- `raw_input` will use `readline` library (which is needed for history and completion) only when stdio or stdout is a terminal [1]. One way to fix this is to pass the output from socket through a pseudo terminal pair.
- `Pdb` is setting `self.raw_input = 0` if a `stdout` argument is passed to it [2].  If `raw_input` is set, then it seems that `Cmd` class is not using `raw_input` and just `read`s from `stdin`.  As we are setting `sys.stdout = stream_from_socket` anyway, probably it's OK to stop passing `stdout` into `Pdb`.
- To make completion and line editing work special characters should not be processed by the terminal where nc/telnet is running.  We need to disable a few terminal features using `stty` to make it work.

So, with this patch: https://gist.github.com/x746e/df338810c022d2ab85fa and if connecting to `ripdb` with:

```
SAVED_STTY=`stty -g`; stty -icanon -opost -echo -echoe -echok -echoctl -echoke; nc 127.0.0.1 4444; stty $SAVED_STTY
```

it looks like completion and line editing are working fine.

[1] https://hg.python.org/cpython/file/2.7/Parser/myreadline.c#l204
[2] https://hg.python.org/cpython/file/2.7/Lib/pdb.py#l65
[3] https://hg.python.org/cpython/file/2.7/Lib/cmd.py#l128

**Comment by lebedov** [comment]
Thanks, @x746e ! Applied your patch and posted a new version.

**Comment by vduseev** [comment]
@x746e Kirill, thank you for your amazing research and input!

@lebedov Lev, thank you for such a useful library.

I'm still having issues with this 4 years after the issue has been resolved. Trying to figure out, why this solution does not work in VS Code PTY on MacOS.

I've tested that a pure (no netcat) running IPDB works just fine (auto-completion included) in the VS Code.
However, no matter which `stty` settings I turn off using `raw`, `-echo`, `-opost`, etc. flags suggested by @x746e I either get a weird output or a command promt that jumps around terminal at each <kbd>Enter</kbd>. But I wasn't able to make the `tab` character be ignored by the line discipline associated with the TTY. I wanted the `tab` sequence to be passed directly to the ipdb via ripdb, so that it is interpreted properly and invokes auto-completion.

Either it is something caused by the BSD implementation of `stty` on macos, or the suggested way of running `ripdb` with auto-completion via netcat is not working.
